Make possible to mark EnableCrushUpdates as false

malayparida2000 opened the pull request:
The enableCrushUpdates field is a bool field, and after we get the storageCluster as a structured object it's impossible to know if the field is set to false because it was not specified or because it was explicitly set to false. Hence till now we were setting the field to true always. 

Now we are getting the storageCluster as unstructured, and we can check if the field exists in the unstructured object. If it does, it means the user has explicitly set the field. If it does not, it means the field is not specified by the user. This way the user can now make the field enableCrushUpdates as false.

Ref-https://issues.redhat.com/browse/DFBUGS-3813

leelavg commented:
jira description explained the golang mechanics but not the business justification, I see the statement `who are upgrading from pre-4.16 scenarios are facing some issues in their setups if they have multiple deviceClasses.` and wondering what `some issues` mean here 🤔.

I believe we all agree this PR is introducing a complexity to fill the gap of incorrect api, to think of alternatives I want to know what `some issues` mean in above para, wrt to provided solution I see below are some of the cons:
1. opens a way to retrofit all the fields which we designed incorrectly
2. unstructured is inherently not type check-able and a change to api need to consider the usage of unstructured code during upgrades, upgrade can get stuck if it's in the path of operator conditions
3. unstructured is afaik not cached (maybe there is an explicit opt in 🤔), so every reconcile does a live lookup and for an object like StorageCluster (which is relatively big) and a reconciler like StorageCluster (which has relatively many watches) this introduces time & memory latencies/burden.

the previous use of unstructured (that I'm aware of) is in provider-server which is laser focused and only do a get this was done as not to import the whole olm repo 😅, so I guess this PR needs more justification.

open to be corrected.

malayparida2000 committed:
Get storageCluster as unstructured to avoid field default confusion

When the storageCluster is fetched and stored as a structured object,
the unspecified fields are set to their default values. Like bool fields
are set to false, string fields are set to empty string, etc.
Due to this we loose the ability to know the user's intent, whether
the bool field is set to false because it was not specified or because
it was explicitly set to false, whether the string field is set to empty
string because it was not specified or because it was explicitly set to
empty string.
If we get the storageCluster as unstructured, we can check if the field
exists in the unstructured object. If it does, it means the user has
explicitly set the field. If it does not, it means the field is not
specified by the user. This way we can know the user's intent and
avoid the confusion of default values.
The client by default does not read unstructured objects from the cache.
So we need to explicitly set the CacheOptions to read unstructured
objects from the cache.

Signed-off-by: Malay Kumar Parida <mparida@redhat.com>

malayparida2000 committed:
Make possible to mark EnableCrushUpdates as false

The enableCrushUpdates field is a bool field, and after we get the
storageCluster as a structured object it's impossible to know if the
field is set to false because it was not specified or because it was
explicitly set to false. Hence till now we were setting the field to
true always.

Now we are getting the storageCluster as unstructured, and we can check
if the field exists in the unstructured object. If it does, it means the
user has explicitly set the field. If it does not, it means the field is
not specified by the user. This way the user can now make the field
enableCrushUpdates as false.

Signed-off-by: Malay Kumar Parida <mparida@redhat.com>

malayparida2000 committed:
Pass some default block pool spec for .mgr, .nfs

Along with the default cephBlockPool we also always create the .mgr
blockPool. Also when nfs is enabled we also create the .nfs blockPool.
We have no way to specify the poolSpecs of these blockPools.
Hence for getting replica-2 working we need to pass the replicated size
spec if specified for the default cephBlockPool to the .mgr & .nfs
blockPools. Also to get multiple deviceClasses working we need to
pass the EnableCrushUpdates & deviceClass if specified for the default
blockPool to the .mgr & .nfs blockPools.

Signed-off-by: Malay Kumar Parida <mparida@redhat.com>

malayparida2000 committed:
Add unit-test changes

Signed-off-by: Malay Kumar Parida <mparida@redhat.com>

malayparida2000 commented on controllers/storagecluster/cephcluster.go:
```diff
@@ -1529,9 +1530,17 @@ func isEncrptionSettingUpdated(clusterWideEncrytion bool, existingDeviceSet []ro
+	// Check if DeviceClass is explicitly set in the unstructured obj else set to the default device class
+	if val, found := getValueAtPath(r.unstructuredSC, append(specifiedPoolSpecPath, "deviceClass")); found {
+		poolSpec.DeviceClass = val.(string)
+	} else if poolSpec.DeviceClass == "" {
```
Okay, Thanks for the clarification removed the check for deviceClass now

malayparida2000 commented:
> jira description explained the golang mechanics but not the business justification, I see the statement `who are upgrading from pre-4.16 scenarios are facing some issues in their setups if they have multiple deviceClasses.` and wondering what `some issues` mean here 🤔.
> 
> I believe we all agree this PR is introducing a complexity to fill the gap of incorrect api, to think of alternatives I want to know what `some issues` mean in above para, wrt to provided solution I see below are some of the cons:
> 
> 1. opens a way to retrofit all the fields which we designed incorrectly
> 2. unstructured is inherently not type check-able and a change to api need to consider the usage of unstructured code during upgrades, upgrade can get stuck if it's in the path of operator conditions
> 3. unstructured is afaik not cached (maybe there is an explicit opt in 🤔), so every reconcile does a live lookup and for an object like StorageCluster (which is relatively big) and a reconciler like StorageCluster (which has relatively many watches) this introduces time & memory latencies/burden.
> 
> the previous use of unstructured (that I'm aware of) is in provider-server which is laser focused and only do a get this was done as not to import the whole olm repo 😅, so I guess this PR needs more justification.
> 
> open to be corrected.

The issues were a long chain of things, starting with the partner setting up clusters with multiple device classes.
The end solution we reached was that they need to disable makeEnableCrushUpdates: false.

Keeping aside this specific customer's issue, we have made all other fields of the poolSpec configurable, except this field, which sounds wrong. So we do need some way for someone to be able to mark EnableCrushUpdates to false.

> 1. opens a way to retrofit all the fields that we designed incorrectly

If rook were to change the type of the api field to a pointer to bool may be that would the cleanest solution. But what I heard was this issue does not affect rook upstream as for them the default is false anyways. So they are unwilling to make an API change for that. And Changing API type introduces risk during upgrade. But open to discuss this more.

>2. upgrade can get stuck if it's in the path of operator conditions

For now the function I added does not return any error, So the reconcile is not blocked due to any problem in parsing the paths in the unstructured object.

>3. unstructured is afaik not cached (maybe there is an explicit opt in 🤔)

You are right. Thanks for letting me know that. I was not aware of it but after looking up the controller runtime package I have now added a client option for caching the unstructured object. Although I have to read more if that would spike the memory usage of the operator pod
